cmd/cosign/cli/fulcio/fulcioverifier/ctutil/ctutil.go

```diff
@@ -0,0 +1,221 @@
+	s, err := ct.NewSignatureVerifier(pubKey)
+	if err != nil {
+		return fmt.Errorf("error creating signature verifier: %w", err)
+	}
+
+	return VerifySCTWithVerifier(s, chain, sct, embedded)
+}
+
+// VerifySCTWithVerifier takes a ct.SignatureVerifier, a certificate chain, and
+// an SCT and verifies whether the SCT is a valid SCT for the certificate at
+// chain[0], signed by the Log whose public key was used to set up the
+// ct.SignatureVerifier.  If the SCT does not verify, an error will be returned.
+//
+// This function can be used with three different types of leaf certificate:
+//   - X.509 Certificate:
+//       If using this function to verify an SCT for a normal X.509 certificate
+//       then it is enough to just provide the end entity certificate in chain.
+//       This case assumes that the SCT being provided is not embedded within
+//       the leaf certificate provided, i.e. the certificate is what was
+//       submitted to the Certificate Transparency Log in order to obtain the
+//       SCT.  For this case, set embedded to false.
+//   - Precertificate:
+//       If using this function to verify an SCT for a precertificate then the
+//       issuing certificate must also be provided in chain.  The precertificate
+//       should be at chain[0], and its issuer at chain[1].  For this case, set
+//       embedded to false.
+//   - X.509 Certificate containing the SCT embedded within it:
+//       If the SCT you wish to verify is embedded within the certificate you
+//       are providing at chain[0], set embedded to true.  VerifySCT will
+//       verify the provided SCT by building the corresponding precertificate.
+//       VerifySCT will return an error if the provided SCT cannot be found
+//       embedded within chain[0].  As with the precertificate case, the issuing
+//       certificate must also be provided in chain.  The certificate containing
+//       the embedded SCT should be at chain[0], and its issuer at chain[1].
+func VerifySCTWithVerifier(sv *ct.SignatureVerifier, chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp, embedded bool) error {
+	if sv == nil {
+		return errors.New("ct.SignatureVerifier is nil")
+	}
+
+	leaf, err := createLeaf(chain, sct, embedded)
+	if err != nil {
+		return err
+	}
+
+	return sv.VerifySCTSignature(*sct, ct.LogEntry{Leaf: *leaf})
+}
+
+func createLeaf(chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp, embedded bool) (*ct.MerkleTreeLeaf, error) {
+	if len(chain) == 0 {
+		return nil, errors.New("chain is empty")
+	}
+	if sct == nil {
+		return nil, errors.New("sct is nil")
+	}
+
+	if embedded {
+		sctPresent, err := ContainsSCT(chain[0], sct)
+		if err != nil {
+			return nil, fmt.Errorf("error checking for SCT in leaf certificate: %w", err)
+		}
+		if !sctPresent {
+			return nil, errors.New("SCT provided is not embedded within leaf certificate")
+		}
+	}
+
+	certType := ct.X509LogEntryType
+	if chain[0].IsPrecertificate() || embedded {
```

Comment:
This file is a copy from the certificate transparency repo. I needed to do this because the package pulls in another package that defines a flag that conflicts with cosignd, so I took the one file we needed.

The library function supports both embedded and precertificate verifications. The verification of both is similar - Remove the poison extension, remove the SCT extension, and validate only the TBS certificate.

Reply:
Oh, if it is a copy, maybe comment at the start of the file? In that case I wouldn't change it, but note that in the `embedded = True` case, that `certType` is not even used, hence why I don't understand why this is part of the condition.

Reply:
Added a comment!